bench/randindex_bench.d

```diff
@@ -0,0 +1,392 @@
+
+Benchmarks for generating ulong from PermutedCongruentialEngine!(xsh_rr, cast(stream_t)2, true):
+randIndexV1: 0.718778 * 10 ^^ 8 calls/s; sum = 21799772797
+randIndexV2: 1.55491 * 10 ^^ 8 calls/s; sum = 21799158450
+
+Benchmarks for generating ulong from PermutedCongruentialEngine!(rxs_m_xs_forward, cast(stream_t)2, true):
+randIndexV1: 0.866363 * 10 ^^ 8 calls/s; sum = 21799794704
+randIndexV2: 2.52525 * 10 ^^ 8 calls/s; sum = 21800197408
+
+//
+// GENERATING UINT BENCHMARKS
+//
+
+Benchmarks for generating uint from MersenneTwisterEngine!(uint, 32LU, 624LU, 397LU, 31LU, 2567483615u, 11LU, 4294967295u, 7LU, 2636928640u, 15LU, 4022730752u, 18LU, 1812433253u):
+randIndexV1: 1.3836 * 10 ^^ 8 calls/s
+randIndexV2: 2.00803 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 2.42571 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from MersenneTwisterEngine!(ulong, 64LU, 312LU, 156LU, 31LU, 13043109905998158313LU, 29LU, 6148914691236517205LU, 17LU, 8202884508482404352LU, 37LU, 18444473444759240704LU, 43LU, 6364136223846793005LU):
+randIndexV1: 1.29997 * 10 ^^ 8 calls/s
+randIndexV2: 1.8269 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 2.30216 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from XorshiftStarEngine!(ulong, 1024u, 31u, 11u, 30u, 11400714819323198483LU, ulong):
+randIndexV1: 2.28637 * 10 ^^ 8 calls/s
+randIndexV2: 3.16081 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 3.70199 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from Xoroshiro128Plus:
+randIndexV1: 2.56082 * 10 ^^ 8 calls/s
+randIndexV2: 4.37158 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 6.9869 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from PermutedCongruentialEngine!(xsh_rr, cast(stream_t)2, true):
+randIndexV1: 1.50404 * 10 ^^ 8 calls/s
+randIndexV2: 2.32221 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 3.86847 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from PermutedCongruentialEngine!(rxs_m_xs_forward, cast(stream_t)2, true):
+randIndexV1: 1.47411 * 10 ^^ 8 calls/s
+randIndexV2: 2.49377 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 4.24178 * 10 ^^ 8 calls/s
+*/
+
+T randIndexV1(T, G)(ref G gen, T m)
+    if(isSaturatedRandomEngine!G && isUnsigned!T)
+{
+    pragma(inline, false);//Try to prevent LDC from doing anything clever with the modulus.
+
+    assert(m, "m must be positive");
+    T ret = void;
+    T val = void;
+    do
+    {
+        val = gen.rand!T;
+        ret = val % m;
+    }
+    while (val - ret > -m);
+    return ret;
+}
+
+version (LDC)
+{
+    //TODO: figure out specific feature flag or CPU versions where 128 bit multiplication works!
+    version (X86_64)
+        private enum bool probablyCanMultiply128 = true;
+    else
+        private enum bool probablyCanMultiply128 = size_t.sizeof >= ulong.sizeof;
```

Comment:
If I am not wrong, LLVM can always handle 128 / 256 bit multiplication, for targets where it is not supported it will be simulated.

Reply:
I thought so too, but travis-ci compilation failed for x86 on Linux.

Reply:
Yep, mir-algorithm uses common mul algo now. But it should be faster then div anyway